# Playground Series - S3E19: Synthetic Dataset Project

## Introduction
This project involves working with a synthetic dataset from the Playground Series. The goal is to explore and apply machine learning techniques to solve the problem using the provided data. The dataset is designed to simulate real-world conditions, providing an opportunity to experiment with various models, feature engineering, and data visualization methods. The data can be downloaded from the [Kaggle competition page](https://www.kaggle.com/competitions/playground-series-s3e19/data).


In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import LabelEncoder


In [4]:
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

In [6]:
train.head()

,id,date,country,store,product,num_sold
0,0,2017-01-01,Argentina,Kaggle Learn,Using LLMs to Improve Your Coding,63
1,1,2017-01-01,Argentina,Kaggle Learn,Using LLMs to Train More LLMs,66
2,2,2017-01-01,Argentina,Kaggle Learn,Using LLMs to Win Friends and Influence People,9
3,3,2017-01-01,Argentina,Kaggle Learn,Using LLMs to Win More Kaggle Competitions,59
4,4,2017-01-01,Argentina,Kaggle Learn,Using LLMs to Write Better,49


In [8]:
train.shape

(136950, 6)

In [10]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136950 entries, 0 to 136949
Data columns (total 6 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   id        136950 non-null  int64 
 1   date      136950 non-null  object
 2   country   136950 non-null  object
 3   store     136950 non-null  object
 4   product   136950 non-null  object
 5   num_sold  136950 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 6.3+ MB


In [12]:
train.isnull().sum()

id          0
date        0
country     0
store       0
product     0
num_sold    0
dtype: int64

In [14]:
combined_data = pd.concat([train, test], ignore_index=True)

In [16]:
le = LabelEncoder()
combined_data['country'] = le.fit_transform(combined_data['country'])
combined_data['store'] = le.fit_transform(combined_data['store'])
combined_data['product'] = le.fit_transform(combined_data['product'])

In [18]:
combined_data['date'] = pd.to_datetime(combined_data['date'])
combined_data['year'] = combined_data['date'].dt.year.astype(int)
combined_data['month'] = combined_data['date'].dt.month.astype(int)
combined_data['day'] = combined_data['date'].dt.day.astype(int)
combined_data.drop(columns=['date'], inplace=True)

In [22]:
train = combined_data[:len(train)]
test = combined_data[len(train):]

In [24]:
X = train.drop(columns=['num_sold'])
y = train['num_sold']

In [28]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

In [30]:
model = RandomForestRegressor(n_estimators=100, random_state=136950)
model.fit(X_train, y_train)

RandomForestRegressor(random_state=136950)

In [31]:
y_pred = model.predict(X_val)

In [32]:
mae = mean_absolute_error(y_val, y_pred)
mae

12.171683096020447

In [44]:
from sklearn.metrics import r2_score

# Predict the target values
y_pred = model.predict(X_val)

# Calculate R-squared
r2 = r2_score(y_val, y_pred)
print(f'R-squared: {r2:.4f}')


R-squared: 0.9845


In [33]:
test_predictions = model.predict(test.drop(columns=['num_sold']))

In [40]:
submission = pd.DataFrame({'id': test['id'], 'num_sold': test_predictions})

submission.to_csv('submission.csv', index=False)

## Conclusion

In this project, the **RandomForestRegressor** model was used to predict continuous values. The model was trained with 100 estimators and evaluated on the validation set. The performance metrics showed promising results:

- The **R-squared (R²)** score of **0.9845** indicates that the model is able to explain approximately 98.45% of the variance in the target variable, which is a strong performance for a regression model.
- The **Mean Absolute Error (MAE)** was **12.17**, meaning that, on average, the model's predictions are off by 12.17 units from the actual values.

Overall, the model performs very well, but further refinements such as hyperparameter tuning or feature engineering could help improve its predictive accuracy even further.
